### ISSUE TITLE
Multidimensional syntax $a[index $c, $d] not supported

### ISSUE THREAD

**Issue by philiprbrenan** [Perl/perl5#18972]
Omitting the parentheses around the arguments to **index**, when used as an array index, produces a misleading error message ```Multidimensional syntax not supported``` which assumes that **index** has only one argument, which, if so, is inconsistent with the preceding ```my $c =    index $a, $b``` which assumes it has two. 

```
  my @a;
  my ($a, $b);
  my $c =    index $a, $b;   # OK
  my $d = $a[index($a, $b)]; # OK
  my $e = $a[index $a, $b];  # Multidimensional syntax $a[index $c, $d] not supported
```
It might be worth noting that the flow analysis performed by **Java** for analyzing assignments to **final** variables would be sufficient to generate warning messages complaining about the incorrect initialization of the variables below:

**Comment by iabyn** [comment]
On Thu, Jul 08, 2021 at 09:23:26PM -0700, philip r brenan wrote:
>   my $e = $a[index $a, $b];  # Multidimensional syntax $a[index $c, $d] not supported

This was fixed in 5.32.0 by this commit:

commit 41eecd54c335a0342b04dbea635695db80579946
Author: Yves Orton ***@***.***>
Date:   Fri Jan 31 15:02:46 2020 +0100

    toke.c: fix Multidimensional array heuristic to ignore function calls
    
    Fix issue #16535 - $t[index $x, $y] should not throw Multidimensional
    array warnings.
    
    The heuristic for detecting lists in array subscripts is implemented
    in toke.c, which means it is not particularly reliable. There are
    lots of ways that code might return a list in an array subscript.
    
    So for instance $t[do{ $x, $y }] should throw a warning but doesn't.
    
    On the other hand, we can make this warning less likely to happen
    by being a touch more careful about how we parse the inside of the
    square brackets so we do not throw an exception from $t[index $x,$y].
    
    Really this should be moved to the parser so we do not need to rely
    on fallable heuristics, and also into the runtime so that if we have
    
        $t[f()]
    
    and f() returns a list we can also warn there. But for now this
    improves things somewhat.


-- 
You're only as old as you look.

**Comment by Grinnz** [comment]
On Fri, Jul 9, 2021 at 7:48 AM iabyn ***@***.***> wrote:

> On Thu, Jul 08, 2021 at 09:23:26PM -0700, philip r brenan wrote:
> > my $e = $a[index $a, $b]; # Multidimensional syntax $a[index $c, $d] not
> supported
>
> This was fixed in 5.32.0 by this commit:
>
> commit 41eecd54c335a0342b04dbea635695db80579946
> Author: Yves Orton ***@***.***>
> Date: Fri Jan 31 15:02:46 2020 +0100
>
> toke.c: fix Multidimensional array heuristic to ignore function calls
>
> Fix issue #16535 - $t[index $x, $y] should not throw Multidimensional
> array warnings.
>
> The heuristic for detecting lists in array subscripts is implemented
> in toke.c, which means it is not particularly reliable. There are
> lots of ways that code might return a list in an array subscript.
>
> So for instance $t[do{ $x, $y }] should throw a warning but doesn't.
>
> On the other hand, we can make this warning less likely to happen
> by being a touch more careful about how we parse the inside of the
> square brackets so we do not throw an exception from $t[index $x,$y].
>
> Really this should be moved to the parser so we do not need to rely
> on fallable heuristics, and also into the runtime so that if we have
>
> $t[f()]
>
> and f() returns a list we can also warn there. But for now this
> improves things somewhat.
>

The last example puts f() in scalar context so it cannot return a list.
Similarly with any other construct not explicitly recognized by the parser
as multidimensional array syntax. This conceptual inconsistency is one of
several reasons the option was added to disable this misfeature.

-Dan

**Comment by tonycoz** [comment]
> This conceptual inconsistency is one of several reasons the option was added to disable this misfeature.

This `$array[$a, $b]` syntax, where a list is supplied as an array subscript, isn't disabled (it simply evaluates the list as the comma operator).

The syntax that can be optionally disabled is where a list is supplied as a hash subscript.

The reported bug has been fixed (and the report really should have included a version number), so closing.
